daml-lf/transaction/src/main/scala/com/daml/lf/TransactionNodesStatistics.scala

```diff
@@ -0,0 +1,148 @@
+    fetchesByCid: Int,
+    fetchesByKey: Int,
+    lookupsByKey: Int,
+    rollbacks: Int,
+) {
+
+  def +(that: TransactionNodesStatistics) =
+    TransactionNodesStatistics(
+      creates = this.creates + that.creates,
+      consumingExercisesByCid = this.consumingExercisesByCid + that.consumingExercisesByCid,
+      nonconsumingExercisesByCid =
+        this.nonconsumingExercisesByCid + that.nonconsumingExercisesByCid,
+      consumingExercisesByKey = this.consumingExercisesByKey + that.consumingExercisesByKey,
+      nonconsumingExercisesByKey =
+        this.nonconsumingExercisesByKey + that.nonconsumingExercisesByKey,
+      fetchesByCid = this.fetchesByCid + that.fetchesByCid,
+      fetchesByKey = this.fetchesByKey + that.fetchesByKey,
+      lookupsByKey = this.lookupsByKey + that.lookupsByKey,
+      rollbacks = this.rollbacks + that.rollbacks,
+    )
+
+  def exercisesByCid: Int = consumingExercisesByCid + nonconsumingExercisesByCid
+  def exercisesByKey: Int = consumingExercisesByKey + nonconsumingExercisesByKey
+  def exercises: Int = exercisesByCid + exercisesByKey
+  def consumingExercises: Int = consumingExercisesByCid + consumingExercisesByKey
+  def nonconsumingExercises: Int = nonconsumingExercisesByCid + nonconsumingExercisesByKey
+  def fetches: Int = fetchesByCid + fetchesByKey
+  def byKeys: Int = exercisesByKey + fetchesByKey + lookupsByKey
+  def actions: Int = creates + exercises + fetches + lookupsByKey
+  def nodes: Int = actions + rollbacks
+}
+
+object TransactionNodesStatistics {
+
+  val Empty = TransactionNodesStatistics(0, 0, 0, 0, 0, 0, 0, 0, 0)
+
+  private[this] val numberOfFields = Empty.productArity
+
+  private[this] val Seq(
+    createsIdx,
+    consumingExercisesByCidIdx,
+    nonconsumingExerciseCidsIdx,
+    consumingExercisesByKeyIdx,
+    nonconsumingExercisesByKeyIdx,
+    fetchesIdx,
+    fetchesByKeyIdx,
+    lookupsByKeyIdx,
+    rollbacksIdx,
+  ) =
+    (0 until numberOfFields)
+
+  private[this] def emptyFields = Array.fill(numberOfFields)(0)
+
+  private[this] def build(stats: Array[Int]) =
+    TransactionNodesStatistics(
+      creates = stats(createsIdx),
+      consumingExercisesByCid = stats(consumingExercisesByCidIdx),
+      nonconsumingExercisesByCid = stats(nonconsumingExerciseCidsIdx),
+      consumingExercisesByKey = stats(consumingExercisesByKeyIdx),
+      nonconsumingExercisesByKey = stats(nonconsumingExercisesByKeyIdx),
+      fetchesByCid = stats(fetchesIdx),
+      fetchesByKey = stats(fetchesByKeyIdx),
+      lookupsByKey = stats(lookupsByKeyIdx),
+      rollbacks = stats(rollbacksIdx),
+    )
+
+  def stats(tx: VersionedTransaction): (TransactionNodesStatistics, TransactionNodesStatistics) =
+    stats(tx.transaction)
+
+  /** This function produces statistics about the "committed" nodes (those nodes
+    *  that do not appear under a rollback node) on the one hand and
+    *  "rollbacked" nodes (those nodes that do appear under a rollback node) on
```

Comment:
say rolled back and then you don't need the quotation marks :-)

Reply:
Good idea.

Reply:
bd6639ae7ba10010c928dea988ed9d375969b692